cmd/erasure-encode.go

```diff
@@ -21,44 +21,36 @@ import (
 	"context"
 	"fmt"
 	"io"
-	"sync"
 )
 
-// Writes in parallel to writers
-type parallelWriter struct {
+// Writes to multiple writers
+type multiWriter struct {
 	writers     []io.Writer
 	writeQuorum int
 	errs        []error
 }
 
-// Write writes data to writers in parallel.
-func (p *parallelWriter) Write(ctx context.Context, blocks [][]byte) error {
-	var wg sync.WaitGroup
-
+// Write writes data to writers.
+func (p *multiWriter) Write(ctx context.Context, blocks [][]byte) error {
 	for i := range p.writers {
-		if p.writers[i] == nil {
-			p.errs[i] = errDiskNotFound
+		if p.errs[i] != nil {
 			continue
 		}
-		if p.errs[i] != nil {
+		if p.writers[i] == nil {
+			p.errs[i] = errDiskNotFound
 			continue
 		}
-		wg.Add(1)
-		go func(i int) {
-			defer wg.Done()
-			var n int
-			n, p.errs[i] = p.writers[i].Write(blocks[i])
-			if p.errs[i] == nil {
-				if n != len(blocks[i]) {
-					p.errs[i] = io.ErrShortWrite
-					p.writers[i] = nil
-				}
-			} else {
+		var n int
+		n, p.errs[i] = p.writers[i].Write(blocks[i])
+		if p.errs[i] == nil {
+			if n != len(blocks[i]) {
+				p.errs[i] = io.ErrShortWrite
 				p.writers[i] = nil
 			}
```

Comment:
are you serializing this on purpose?

Reply:
@klauspost ^^

Reply:
@harshavardhana It is pointless to spin up goroutines just for mem-copies. As I wrote:

> Since there is always "space" to write output data we remove the parallel writing, and we are always writing to memory now and the goroutine synchronization overhead probably isn't worth taking. If output was blocked in the existing, we would still wait for it to unblock in parallelWriter, so it would make no difference there - except now the ringbuffer smoothes out the load.

Reply:
(we are either writing to a ringbuffer (to disk/net) or a byte.Buffer (inline).